pytorch_lightning/trainer/supporters.py

```diff
@@ -120,23 +130,126 @@ def mean(self):
 
 
 class PredictionCollection(object):
-    def __init__(self, global_rank: int, world_size: int):
+    def __init__(self, global_rank: int, world_size: int, all_gather: Callable):
         self.global_rank = global_rank
         self.world_size = world_size
-        self.predictions = {}
-        self.num_predictions = 0
+        self.all_gather = all_gather
+        self._legacy_predictions = {}
+        self._predictions = {stage: {} for stage in LoggerStages}
+
+    @property
+    def predictions(self):
+        return self._predictions[self.current_stage]
+
+    @property
+    def should_all_gather(self):
+        return torch.distributed.is_available() or not torch.distributed.is_initialized() and self.world_size > 1
+
+    @staticmethod
+    def convert_to_numpy(value):
+        return value.cpu().numpy()
+
+    def _cache_prediction(self, predictions, dl_idx: int):
+        internal_predictions = self.predictions
+
+        if dl_idx not in internal_predictions:
+            internal_predictions[dl_idx] = {}
+
+        for pred in predictions:
+            if "path" in pred:
+                if self.should_all_gather:
+                    raise MisconfigurationException(
+                        "`path` key is not supported yet in distributed setting. Please, use `id`. "
+                    )
+                key = pred["path"]
+
+            elif "id" in pred:
+                key = pred["id"]
+                if not isinstance(key, (int, float, torch.Tensor)):
```

Comment:
is it a good idea to allow `torch.Tensor` which can be a multidimensional matrix...

Reply:
Good catch ! Removed !